fix(numpy-param-doc): allow "default" in params

adam-grant-hendry opened the pull request:
- [x] Write a good description on what the PR does.
- [x] Create a news fragment with `towncrier create <IssueNumber>.<type>` which will be
  included in the changelog. `<type>` can be one of: new_check, removed_check, extension,
  false_positive, false_negative, bugfix, other, internal. If necessary you can write
  details or offer examples on how the new change is supposed to work.
- [x] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

NumPy doc style guide permits using `default` in parameter signatures.

Closes ##6211

adam-grant-hendry committed:
fix(numpy-param-doc): allow "default" in params

NumPy doc style guide permits using `default` in parameter
signatures.

Fixes: #6211

adam-grant-hendry commented:
@DanielNoord @Pierre-Sassoulas Would you please kindly help review? Thank you.

Pierre-Sassoulas commented on tests/functional/ext/docparams/docparams.py:
```diff
@@ -92,3 +93,40 @@ def params_are_documented(par1: int, *, par2: int) -> int:
+    print(arg1, arg2)
+
+
+def regression_6211(x: int = 0) -> None:
+    """This is a regression test for issue #6211.
+
+    False negative of "missing param doc" was being issued when "default" used in
+    NumPy-style docs. This test should return no errors.
+
+    See https://github.com/PyCQA/pylint/issues/6211
+
+    Parameter
+    ---------
+    x : int, default 0
+        The x parameter
+    """
+
+    print(x)
```
```suggestion
def regression_6211(number: int = 0) -> None:
    """This is a regression test for issue #6211.

    False negative of "missing param doc" was being issued when "default" used in
    NumPy-style docs. This test should return no errors.

    See https://github.com/PyCQA/pylint/issues/6211

    Parameter
    ---------
    number : int, default 0
        The number parameter
    """

    print(number)
```

adam-grant-hendry replied:
Good catch. This was causing an `invalid-name` error. I will fix.

Pierre-Sassoulas replied:
You can commit the suggestion directly in interface :)

Pierre-Sassoulas reviewed (changes requested):
Could you add a fragment for the changelog please ? :)

DanielNoord commented:
@adam-grant-hendry The primer tests shows that we would be emitting a lot more messages which I don't really expect. Perhaps we can take a test from there?

Pierre-Sassoulas committed:
Bump pylint to 2.15.0, update changelog (#7355)

* Add Marc Byrne to the maintainer team

adam-grant-hendry committed:
fix(formal-param): Change `x` to `number`

Change `x` to `number` in test method `regression_6211`. This was
causing an `invalid-name` error.

adam-grant-hendry commented:
> @adam-grant-hendry The primer tests shows that we would be emitting a lot more messages which I don't really expect. Perhaps we can take a test from there?

One difficulty I have is that `pytest` isn't outputting what the regular expressions are matching and not matching. Is there a way I can do this with the `pylint` source, or do I have to write a throw-away script using `re` and check for matches?

adam-grant-hendry committed:
feat(news): add towncrier fragment

Adds a news fragment for the issue.

adam-grant-hendry committed:
feat(news): add towncrier fragment

Adds a news fragment for the issue.

adam-grant-hendry commented:
> Could you add a fragment for the changelog please ? :)

Yes, I added a fragment with my last push.

adam-grant-hendry committed:
fix(news): correct fragment to #6211

Fix was for issue #6211. The PR that fixes this is #7360.

See #6211

Pierre-Sassoulas reviewed (commented):
Would you mind rebasing/merging the main branch ? (there's modification from the recent release mixed in the diff.)

Without looking into it, I think something is wrong with the modified regex as the primer raise a lot more message now.

adam-grant-hendry commented:
> Would you mind rebasing/merging the main branch ? (there's modification from the recent release mixed in the diff.)

Just did, but couldn't figure out how to do that without force pushing. I used `--force-with-lease` to try not to overwrite work you or @DanielNoord may have done.
 
> Without looking into it, I think something is wrong with the modified regex as the primer raise a lot more message now.

Yes, the regex is not working, but I don't know why.

adam-grant-hendry committed:
fix(regex): fix `re_only_desc` and add lookaheads

The `re_only_desc` regex did not match for white and characters after
`\n`, so some description-only lines weren't getting matched. In
addition, lookaheads were added to `re_param_line` (i.e. make sure the
type group is not followed by a new line (`\n`)). Lastly, named groups
(ala Perl regular expressions) were added for slightly improved clarity.

The hyperlink in the body of test `regression_6211` was causing
problems, so this was moved into a reStructuredText directive.

Finally, the added

    `from __future__ import annotations`

moved all lines in

	`tests/functional/ext/docparams/docparams.py`

down one line, so the line numbers in `docparams.txt` were off by 1,
causing tests to fail. This was corrected.

adam-grant-hendry commented:
@DanielNoord @Pierre-Sassoulas This should be working now. I wrote a test script to help me out:

<details>
<summary><b>test_script.py</b></summary>

```python
from __future__ import annotations

import re
import typing

from pylint.extensions._check_docs_utils import NumpyDocstring

if typing.TYPE_CHECKING:
    from pylint.extensions._check_docs_utils import Docstring


def space_indentation(s: str) -> int:
    """The number of leading spaces in a string.

    :param str s: input string

    :rtype: int
    :return: number of leading spaces
    """
    return len(s) - len(s.lstrip(" "))


def _is_section_header(line: str) -> bool:
    return bool(re.match(r"\s*-+$", line))


def min_section_indent(section_match: re.Match[str]) -> int:
    return len(section_match.group(1))


def _parse_section(
    section_re: re.Pattern[str],
    doc: str,
) -> list[str]:
    section_match = section_re.search(doc)
    if section_match is None:
        return []

    min_indentation = min_section_indent(section_match)

    entries: list[str] = []
    entry: list[str] = []
    is_first = True
    for line in section_match.group(2).splitlines():
        if not line.strip():
            continue
        indentation = space_indentation(line)
        if indentation < min_indentation:
            break

        # The first line after the header defines the minimum
        # indentation.
        if is_first:
            min_indentation = indentation
            is_first = False

        if indentation == min_indentation:
            if _is_section_header(line):
                break
            # Lines with minimum indentation must contain the beginning
            # of a new parameter documentation.
            if entry:
                entries.append("\n".join(entry))
                entry = []

        entry.append(line)

    if entry:
        entries.append("\n".join(entry))

    return entries


def match_param_docs(
    checker: NumpyDocstring,
    doc: str,
) -> tuple[set[str], set[str]]:
    """Matches parameter documentation section to parameter documentation rules."""
    params_with_doc = set()
    params_with_type = set()

    entries = _parse_section(checker.re_param_section, doc)
    entries.extend(_parse_section(checker.re_keyword_param_section, doc))
    for entry in entries:
        match = checker.re_param_line.match(entry)
        if not match:
            continue

        print(match)
        print()
        # check if parameter has description only
        re_only_desc = re.match(r"\s*(\*{0,2}\w+)\s*:?\n\s*\w*$", entry)
        print(f're_only_desc: {re_only_desc}')
        if re_only_desc:
            param_name = match.group('param_name')
            param_desc = match.group('param_type')
            param_type = None
        else:
            param_name = match.group('param_name')
            param_type = match.group('param_type')
            param_desc = match.group('param_desc')

        print(f"Param Name: '{param_name}'")
        print(f"Type: '{param_type}'")
        print(f"Description: '{param_desc}'")

        if param_type:
            params_with_type.add(param_name)

        if param_desc:
            params_with_doc.add(param_name)

    return params_with_doc, params_with_type


if __name__ == '__main__':
    docstring_1 = """This is a regression test for issue `#6211`_.

False negative of "missing param doc" was being issued when "default" used in
NumPy-style docs. This test should return no errors.

Parameters
----------
number : int, default 0
    The number parameter

.. _`#6211`:
    https://github.com/PyCQA/pylint/issues/6211
"""

    docstring_2 = """No errors raised when pipe symbol used for or.

`PEP 604`_ allows writing Union types as X | Y. Can be enabled in Python <3.10
using `from __future__ import annotations`.

Parameters
----------
arg1
    The first arg
arg2 : str | None, default=None
    The second arg

.. _`PEP 604`:
    https://peps.python.org/pep-0604/
"""

    docstring_3 = """Reported in https://github.com/PyCQA/pylint/issues/4035.

Parameters
----------
arg : int
    The number of times to print it.
option : {"y", "n"}
    Do I do it?
option2 : {"y", None, "n"}
    Do I do it?

"""

    docstring_4 = """function foobar ...

Parameters
----------
arg1 : int
arg2 : int
arg3 : str
"""

    docstring_5 = """function foobar ...
Parameters
----------
arg1:
    description
arg2: str
"""

    np_checker = NumpyDocstring(None)

    docstrings = [
        docstring_1,
        docstring_2,
        docstring_3,
        docstring_4,
        docstring_5,
    ]

    for ndx, docstr in enumerate(docstrings):
        print(f'========= Docstring {ndx+1} =========')
        params_with_description_only, params_with_a_type = match_param_docs(np_checker, docstr)

        print()
        print(f'Parameters with a description only: {params_with_description_only}')
        print(f'Parameters with a type: {params_with_a_type}')
        print()

```
</details>

A couple notes:

1. The OP to Issue #6211 used "Parameter" (singular) instead of the intended "Parameters" (plural), and as a result this regex fails to find the parameters section:

```python
re_param_section = re.compile(
        _re_section_template.format(r"(?:Args|Arguments|Parameters)"),
        re.X | re.S | re.M,
    )
```

2. The hyperlink I put in the body of the `regression_6211` docstring was causing errors (I suspect because of the colon after the "https"). I turned this into a directive.

3. Adding the `from __future__ import annotations` line to the top of `docparams.py` (so I can use python `3.10` style type annotations) shifted each function down by one line, so I needed to update the `docparams.txt` file.

4. The regular expression for default values only accounts for strings, booleans, and `None`, but not:

 - numbers (ints (`2`), floats (`2.35`), binary (`0b010`), oct (`0o007`), or hex (`0xDEADBEEF`)
 - those who use `_` as separators between sets of 3 digits (e.g. `10_000_000`)
 -  byte strings (i.e. `b'this will come to byte you')
 - callables (i.e. `my_funct=jerry`)

There may be other literal/default types to add in the future, but for now I accounted just for ints because that solves the issue at hand:

```python
re_default_value = r"""((['"]\w+\s*['"])|(\d+)|(True)|(False)|(None))"""
```

adam-grant-hendry committed:
fix(pre-commit): fix pre-commit errors

Also remove lookahead from `re_param_line` as tests pass without it.

adam-grant-hendry commented:
> @adam-grant-hendry The primer tests shows that we would be emitting a lot more messages which I don't really expect. Perhaps we can take a test from there?

@DanielNoord @Pierre-Sassoulas After using my script, I found the original code was matching colon (`:`) for `type` instead of the
actual type. Most of these errors are accurate:

    - 1: `pandas/compat/pickle_compat.py`#L256
	- 6: `pandas/tests/extension/conftest.py`#L47

have the form

```
param : description instead of type
```

This is incorrect per the [NumPy Style Guide](https://numpydoc.readthedocs.io/en/latest/format.html#parameters). Descriptions must go
on the next line indented. Only types can occur after the colon on the same line.

    - 7, 8: `pandas/tests/io/test_parquet.py#L165`
	
is in fact missing documentation for the listed parameters.

    - 9, 13: `pandas/tests/io/formats/test_to_html.py#L30` (there are many others)

has incorrect types. `"pytest fixture"` and `matplotlib Axes object` are not types; these should be a class name (dotted attribute), type alias, or builtin (e.g. `_pytest.monkeypatch.MonkeyPatch` or `matplotlib.axes`)


If `pandas` wishes to use a style different from `numpy`, they should have their own style guide. It is confusing for those wanting to follow
the `numpy` guide to get unexpected allowed behavior (and it might cause `sphinx` not to work properly). For this, I recommend the `pylint`
team create a separate class in `_check_docs_utils.py` and call it `PandasDocstring` and to have the `pandas` team confirm it.

Numbers 2-# (from `pandas/tests/arithmetic/common.py`) have method formal parameters that are not type hinted, and hence should technically be
`Any`.

DanielNoord commented:
@adam-grant-hendry That does indeed seem like we were incorrectly matching previously.

Would you be willing to open a separate PR to fix that and add some regressions tests against that?

adam-grant-hendry commented:
> @adam-grant-hendry That does indeed seem like we were incorrectly matching previously.
> 
> Would you be willing to open a separate PR to fix that and add some regressions tests against that?

I would, but I'm not sure I have time at this stage. I doubt I would be able to finish this PR even. The amount of fixes required is more than I originally envisioned, I'm sorry.

At this stage, I'm simply getting by by:

1. Not using `default` in the type, but in the description (e.g. `Defaults to ...`)
2. Using `or` and `of` instead of `|` and `[]`, respectively (though this is actually per the `numpydoc style guide` and the latter is from the newer python typing annotations syntax)

adam-grant-hendry commented:
@DanielNoord @Pierre-Sassoulas Two more drive-by comments:

1. Because `:` is searched for in the regex, this also causes false-negatives when directives are used in the docstring body. Any colons in the body is going to be an issue.

e.g. This gives a `missing-param-doc` error

```python
def func(self, a: int, b: bool) -> None:
    """Does something funky.

    This is where you explain the funk.

    ..warning:: `a` Explodes!
        Don't use `a`

    Parameters
    ------------
    a : int
        An integer
    b : bool
        A bool
```

2. Your regex's don't match when a union of types ends with `, or type`. It matches `type of type` and `type or type`, but the following does not work:

```python
def funkify(self, a: int | bool | str) -> str:
    """Get with the funk.

    What the funk?!

    Parameters
    ------------
    a : int, bool, or str
        This is an ``int``, ``bool``, or ``str``. How funky!

    Returns
    --------
    b : str
        Because why not return a ``str``
```

DanielNoord committed:
Some fixes

DanielNoord committed:
Fix output

DanielNoord commented:
> @DanielNoord @Pierre-Sassoulas Two more drive-by comments:

Both of these should probably be their own issue, right?

I have updated this PR after the merge of #7398. Let's see what the primer thinks!

DanielNoord commented:
@Pierre-Sassoulas We have a similar issue as discussed previously here.

Should we just accept anything that comes on the `type` line in a `Numpy` docstring? It feels better to do so than to try and determine whether `type or None` or `a description of a type` is a "real type".

Pierre-Sassoulas committed:
Merge branch 'main' into fix/numpy-param-doc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Pierre-Sassoulas reviewed (commented):
Sorry for the delay I did not realize this was waiting for my input. I think we should accept anything from the type field, because being able to tell valid type from text is hard, and doing something like "None or MyClass" is frequent enough that the primers message was truncated because there's so much new message. (Removing from the 3.0 milestone because this is the kind of thing that can be merged later)

Pierre-Sassoulas committed:
Be more permissive

Pierre-Sassoulas committed:
Merge branch 'main' into takeover-7360

Pierre-Sassoulas committed:
Fix line numbers in docparams.txt after merge

The merge of main into the PR branch shifted line numbers in
docparams.py by +2 (due to a blank line and the
`# pylint: disable=broad-exception-raised` comment added at the top
on main). Regenerate the expected output accordingly.

Pierre-Sassoulas committed:
Handle NumPy multi-param-per-line idiom

Primer surfaced false positives for `a, b : array-like` and
`top, bottom, left, right : bool, optional`, where NumPy style
documents several parameters of the same type on a single line.
The regex now accepts comma-separated names and `match_param_docs`
splits them so the type and description apply to each.

DanielNoord reviewed (approved):
Let's merge this :)
